scanner/src/main/kotlin/storages/CompositeStorage.kt

```diff
@@ -0,0 +1,115 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.scanner.storages
+
+import org.ossreviewtoolkit.model.Failure
+import org.ossreviewtoolkit.model.Identifier
+import org.ossreviewtoolkit.model.Package
+import org.ossreviewtoolkit.model.Result
+import org.ossreviewtoolkit.model.ScanResult
+import org.ossreviewtoolkit.model.ScanResultContainer
+import org.ossreviewtoolkit.model.ScannerDetails
+import org.ossreviewtoolkit.model.Success
+import org.ossreviewtoolkit.scanner.ScanResultsStorage
+
+/**
+ * A [ScanResultsStorage] implementation that manages multiple concrete storages for reading and writing scan results.
+ *
+ * This class can be used to construct a flexible setup of storages to use to obtain scan results requested for the
+ * current run and to persist them to speed up later runs. When a scan result is requested the class iterates
+ * over the list of configured reader storages and returns the first successful, non-empty result. So a prioritized
+ * chain of places where to look up results can be configured.
+ *
+ * Analogously, when a result is to be written the class passes the result to all the storages configured as writers.
+ * For many use cases, a single writer is probably sufficient; but if there are special requirements - e.g. to
+ * create backups of results automatically or to have a fast, but transient cache of results -, it is possible to
+ * specify multiple writers.
+ */
+class CompositeStorage(
+    val readers: List<ScanResultsStorage>,
+    val writers: List<ScanResultsStorage>
+) : ScanResultsStorage() {
+    override val name: String
+        get() {
+            val readerNames = readers.map { it.name }
+            val writerNames = writers.map { it.name }
+            return readerNames.joinToString(separator = ",", prefix = "composite[") +
+                    writerNames.joinToString(separator = ",", prefix = "/", postfix = "]")
+        }
+
+    /**
+     * Try to find scan results for the provided [id]. This implementation iterates over all the reader storages
+     * provided until it receives a non-empty success result.
+     */
+    override fun readFromStorage(id: Identifier): Result<ScanResultContainer> =
+        fetchReadResult(id) { read(id) }
+
+    /**
+     * Try to find scan results for the provided [pkg] and [scannerDetails]. This implementation iterates over all
+     * the reader storages provided until it receives a non-empty success result.
+     */
+    override fun readFromStorage(pkg: Package, scannerDetails: ScannerDetails): Result<ScanResultContainer> =
+        fetchReadResult(pkg.id) { read(pkg, scannerDetails) }
+
+    /**
+     * Trigger all configured writer storages to add the [scanResult] for the given [id]. Return a success result
```

Comment:
What's the rational for supporting multiple writers?

I'm asking since replication setup could also be considered out-of-scope of ORT.

Reply:
The original intent was only to support multiple readers, but since we do not formally separate readers and writers but only have a common `ScanResultsStorage` class, it was rather easy and consistent to also allow multiple writers. Having that feature could be beneficial e.g. for users who want to "cache" slower storage like ClearlyDefined in faster local file storage, or to be more independent of third-party infrastructure.

Reply:
To me it at least smells like unneeded complexity. I find the semantics of the function / contract becomes quite a bit more complex. Anyhow not going to block it.